Step4 로또(2등) 코드 리뷰 요청드립니다.

dongha-byun opened the pull request:
안녕하세요! 
우아한 테크캠프 Pro 5기 교육생 변동하 입니다.

4단계 로또(2등) 미션 코드리뷰를 요청드립니다!

추가로 지난 피드백에서 말씀해주신 코드컨벤션 스타일은 제 개발환경에선 적용된 것 같은데 깃 상에서 표현이 안되는 거 같습니다...
혹시 몰라서 적용시킨 IDE 화면 첨부해드립니다.
<img width="1438" alt="스크린샷 2022-11-02 오후 2 17 09" src="https://user-images.githubusercontent.com/95368245/199569128-f10f85b6-1589-4ee1-8a10-17d455b57d41.png">

감사합니다!

dongha-byun committed:
Refactor: 패키지명 수정 (lottoauto -> lotto)

dongha-byun committed:
Docs: 로또 2등 요구사항 및 기능목록 정의

dongha-byun committed:
Remove: 미사용 class 및 테스트 삭제

dongha-byun committed:
Test: 구매로또 컬렉션 class 테스트코드 작성

dongha-byun committed:
Feat: 구매로또 컬렉션 class 구현코드 작성

dongha-byun committed:
Test: 테스트코드 내에서 사용할 LottoNumberStrategy interface 구현체 추가

dongha-byun committed:
Test: 당첨로또 class 테스트코드 작성

dongha-byun committed:
Feat: 당첨로또 구현 class 추가
 - 기본 6개 번호와 보너스 로또 번호 구분을 위한 enum 포함

dongha-byun committed:
Refactor: 코드 리펙터링
 - LottoPayment static factory method 추가

dongha-byun committed:
Refactor: 코드 리펙터링
 - Lotto 당첨 결과 enum 에서 관리하도록 수정
 - LottoResult 내 컬렉션에 key 타입 변경( Integer -> LottoWinningMoneyEnum )

dongha-byun committed:
Refactor: 코드 리펙터링
 - service layer 축소 ( 기존 3개 -> 1개 )
 - Lotto class equals, hashCode 메서드 재정의

dongha-byun committed:
Refactor: 코드 리펙터링
 - 입출력 영역 수정
 - main 메서드 수정

chae-yh commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,6 +1,7 @@
-package lottoauto.domain;
+package lotto.domain;
```
적절하게 패키지명이 변경되었네요 👍

chae-yh commented on src/main/java/lotto/README.md:
```diff
@@ -41,3 +41,21 @@
+### 2) 기존번호 5개 + 보너스 번호 1개 가 맞을 경우, 2등으로 처리한다.
+#### 2-1) 2등 당첨금은 3,000,000 원 이다.
+
+## 2. 기능 목록 정리
```
기능 요구 사항 정리 👍

chae-yh commented on src/main/java/lotto/domain/BuyingLottoGroup.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public List<Lotto> getLottos() {
+        return lottos;
```
이렇게 되면 외부에서 해당 값을 쉽게 변경해줄 수 있는데, [방어적 복사](https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/)를 사용하면 어떨까요...?

dongha-byun replied:
생각 못한 부분이었습니다...! 피드백 감사합니당!!

chae-yh commented on src/main/java/lotto/domain/BuyingLottoGroup.java:
```diff
@@ -0,0 +1,30 @@
+import lotto.strategy.LottoNumberStrategy;
+
+public class BuyingLottoGroup {
+    private List<Lotto> lottos = new ArrayList<>();
```
일급 컬렉션 👍 다만, 해당 형태는 값이 mutable 하네요. [불변으로 바꿔줄 수](https://jojoldu.tistory.com/412)도 있을 것 같습니다...!

dongha-byun replied:
넵...! 이부분도 고쳐보겠습니다!

chae-yh commented on src/main/java/lotto/domain/LottoPayment.java:
```diff
@@ -18,7 +22,7 @@ public void validateOnlyNumber(String payment) {
     }
 
-    public int buyLotto() {
+    public int getBuyLottoCount() {
```
메서드명을 적절하게 동사로 바꿔주셨군요 👍 
get과 buy가 둘 다 동사라, getLottoCount 정도로만 해도 좋을 것 같습니다

dongha-byun replied:
아 그러네요 ㅠㅠ 이름 짓는데 좀 더 시간을 투자헤봐야겠습니다 ㅠㅠ 좋은 피드백 감사드립니다!!

chae-yh commented on src/main/java/lotto/domain/LottoWinningMoneyEnum.java:
```diff
@@ -0,0 +1,38 @@
+        this.winningMoney = winningMoney;
+    }
+
+    public static LottoWinningMoneyEnum findEnumByContainCountAndBonusContain(int containCount,
```
findEnumByContainCountAndBonusContain 라는 메서드명에서 findEnum이 되니, 정확히 어떤 값을 가져오는지 확인이 어려운 것 같습니다.

dongha-byun replied:
넵! 메서드 명칭도 다시 고민해보겠습니다!

chae-yh commented on src/main/java/lotto/domain/LottoWinningMoneyEnum.java:
```diff
@@ -0,0 +1,38 @@
+    FOURTH(4, false,50000),
+    THIRD(5, false,1500000),
+    SECOND(5, true, 3000000),
+    FIRST(6, false,2000000000);
```
이런 부분에서 코드 템플릿이 잡아 줄텐데, 아마 템플릿 적용 전의 값들은 그대로 남아 있는 것 같군요...!

dongha-byun replied:
그런가봐요...ㅠㅠ 저한테는 이렇게 표시됩니다! 
![스크린샷 2022-11-04 오후 9 53 31](https://user-images.githubusercontent.com/95368245/199977209-c917ef1e-e41e-4e47-8343-c4223941f809.png)

chae-yh commented on src/main/java/lotto/domain/LottoWinningMoneyEnum.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.Arrays;
+
+public enum LottoWinningMoneyEnum {
```
이미 클래스가 enum으로 명시가 되어 있는데, 클래스명에도 Enum을 명시해줘야 할까요...?

dongha-byun replied:
아 기존에 있던 LottoWinningMoney class 를 삭제하고 여기 이름도 변경했어야 했는데 누락했던 것 같습니다 ㅠㅠ 적절한 이름으로 수정해서 커밋하겠습니다!

chae-yh commented on src/test/java/lotto/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,93 @@
+import static org.assertj.core.api.Assertions.*;
+
+public class LottoResultTest {
+    WinningLotto winningLotto;
```
해당 형태면 접근제한자가 protected로 걸릴텐데, 명시해서 작성(private / protected 등)하면 좋을 것 같습니다.

dongha-byun replied:
네! 앞으론 테스트코드 작성 시에도 신경쓰겠습니다! 감사합니다!!

chae-yh commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,40 @@
+
+public class WinningLotto {
+
+    private Map<WinningLottoType, Object> numbers = new HashMap<>();
```
Lotto와 LottoNumber를 필드로 가지고 있으면 될 것 같은데, 혹시 Map으로 작성하신 이유 여쭤봐도 될까요...?

dongha-byun replied:
일급컬렉션으로 써볼 수 있을까 생각해보다가 Map 으로 한 번 써봤습니다. 
value 가 Object인게 좀 걸려서 고민하다가 한번 이렇게 해보자! 했는데 역시 어색한가 봅니다 ㅠㅠ
Map 으로 해봤으니 Lotto 와 LottoNumber 를 필드로 가지게끔 수정도 해보겠습니다!

chae-yh commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -46,4 +47,21 @@ public String toString() {
     }
+
+    @Override
+    public boolean equals(Object o) {
```
이 부분은 궁금해서 질문드립니다. equals와 hashCode를 재정의하신 이유가 있을까요...!

dongha-byun replied:
WinngingLotto 테스트코드 작성할 때 필요해서 만들었는데, 해당 테스트코드가 중복되는 부분이 있어서 지워놓고 여기는 남겨둔 거 같습니다...

chae-yh commented on src/test/java/lotto/strategy/TestLottoNumberStrategy.java:
```diff
@@ -0,0 +1,11 @@
+public class TestLottoNumberStrategy implements LottoNumberStrategy{
+    @Override
+    public List<Integer> generateNumbers() {
+        return Arrays.asList(1,2,3,4,5,6);
```
이렇게 되니 테스트 fixture 같네요 ㅎㅎ

dongha-byun replied:
BuyingLottoGroupTest 와 WinningLottoTest 두 군데서 사용돼서 별도의 파일로 생성했는데, 테스트 코드 작성 시 보통 이렇게도 쓰나요...? 실무에서 작성해본 적이 없어서 관행(?) 이나 좋은 예시를 잘 모르겠습니다..ㅠ

chae-yh replied:
넵 이런식으로 구현체를 Mocking해서 사용하는 경우가 있습니다. 저는 괜찮은 것 같아요 ㅎㅎ

chae-yh commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,40 @@
+        return lotto.getContainNumberCount((Lotto) numbers.get(BASIC));
+    }
+
+    public boolean isContainBonusNumber(Lotto lotto) {
```
getMatchCountByBasic / isContainBonusNumber 는 private으로 막아줄 수 있지 않을까요...?

dongha-byun replied:
두 메서드에 대한 테스트코드가 있긴한데 비지니스 로직 내에서 해당 객체에서만 쓰이면 별도로 작성하지 않아도 되나요?

chae-yh replied:
private 메서드의 경우 결국 public으로 열린 메서드에서 사용될 것이기 때문에, 해당 public 메서드가 정상 동작하는 것으로 private 메서드도 정상 동작하는 것으로 판단할 수 있을 것 같습니다...!

물론 실무에서는 private 메서드도 테스트하는 경우가 있지만, 저희가 학습하는 개념하에서는 private 메서드에 대해서 테스트는 필요치 않을 것 같아요

chae-yh commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,42 @@
+    @BeforeEach
+    void beforeEach(){
+        winningLotto = create(
+                Lotto.create(Arrays.asList(1,2,3,4,5,6)), LottoNumber.create(7)
```
혹시 당첨 번호가 1, 2, 3, 4, 5, 6 이고 보너스 번호가 6일 경우에는 어떻게 동작해야 할까요...?

dongha-byun replied:
기본적인 검증 로직인데 코드로 구현을 안했네요...ㅜㅜ 이 부분은 적절하게 수정해보도록 하겠습니다!

chae-yh reviewed (changes requested):
동하님 안녕하세요...! 피드백이 조금 늦었네요...ㅠ
우선 step3의 피드백까지 반영하신 상태에서 코드 작성 주셨는데, 전체적으로 잘 작성되어서, 조금만 더 고민해주시면 좋을만한 내용을 코멘트 드렸으니 확인 부탁드립니다 ㅎㅎ

감사합니다 : )

dongha-byun committed:
Refactor: BuyingLottoGroup 일급컬렉션 수정
- 필드변수 lottos 불변 처리 및 방어적 복사 처리
- 테스트 코드 수정

dongha-byun committed:
Refactor: 메서드명 수정
- LottoPayment class 내 메서드명 변경 (getBuyLottoCount -> getLottoCount)
- 테스트코드 및 관련 구현코드 수정

dongha-byun committed:
Refactor: enum class 명 변경

dongha-byun committed:
Refactor: class 및 메서드 명 변경
- LottoWinningMoney -> LottoWinningRank enum 명 변경
- LottoWinningRank 메서드 명 변경 ( findEnumByContainCountAndBonusContain -> findRankByContainCountAndBonusContain )

dongha-byun committed:
Refactor: WinningLotto class refactoring

dongha-byun committed:
Refactor: WinningLotto class refactoring
- WinningLotto 내부에서만 사용하는 메서드 private 처리( 관련 테스트코드 삭제_
- WinningLotto class 필드변수 변경으로 인한 미사용 enum 제거 (WinningLottoType)

dongha-byun committed:
Refactor: WinningLotto class Refactoring

- 보너스볼 중복 검증 validateDuplicateBonusNumber 메서드 추가

dongha-byun committed:
Fix: Lotto class 검증 오류 수정
- Lotto 내 LottoNumber 중복 체크 오류 수정

dongha-byun committed:
Refactor: 미사용 코드 제거
- Lotto class 내 equals / hashcode 미사용으로 인해 제거

chae-yh reviewed (approved):
동하님 안녕하세요. 피드백 잘 반영해주셨습니다 👍 몇가지 코멘트는 달았으나, 사소한 내용들이라 살짝 확인만 부탁드립니다 ㅎㅎ

이제 마지막 step이네요...! 마지막까지 파이팅입니다 : )
